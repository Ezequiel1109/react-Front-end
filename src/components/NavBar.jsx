import { Link, useLocation, useNavigate } from "react-router-dom";
import { useSelector, useDispatch } from "react-redux";
import { logout } from "../services/UserService";
import { selectIsAuth, selectUsername } from "../features/selectors";

export const Navbar = () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();

  //selectores memoizados
  const isAuth = useSelector(selectIsAuth);
  const userMemoized = useSelector(selectUsername);

  const handleLogout = () => {
    dispatch(logout());
    navigate("/login");
  };

  return (
    <nav className="navbar navbar-expand-lg navbar-dark bg-dark mb-4">
      <div className="container">
        <Link className="navbar-brand" to="/">
          Mi App Productos
        </Link>
        <div className="navbar-nav ms-auto">
          {isAuth ? (
            <div className="user-menu">
              <span className="navbar-text me-3">
                Bienvenido, {userMemoized}
              </span>
              <Link
                to="/products"
                className="nav-link">Productos</Link>
              <button onClick={handleLogout} className="nav-link logout-btn">
                Cerrar Sesi√≥n
              </button>
            </div>
          ) :  (
            <>
              <Link className="nav-link" to="/login">Login</Link>
              <Link className="nav-link" to="/register">Registro</Link>
            </>
          )}
        </div>
      </div>
    </nav>
  );
};
