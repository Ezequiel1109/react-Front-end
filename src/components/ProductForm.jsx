import { useEffect, useState } from "react";

const initialDataForm = {
  id: 0,
  name: "",
  price: "",
  quantity: "",
  description: ""
};
// eslint-disable-next-line react/prop-types
export const ProductForm = ({ productSelected, handlerAdd }) => {
  const [form, setForm] = useState(initialDataForm);
  const {id, name, price, quantity, description } = form;

  useEffect(() => {
    setForm(productSelected);
  }, [productSelected]);

  return (
    <form
      onSubmit={(event) => {
        event.preventDefault();
        if (!name || !price || !quantity || !description) {
          alert("debe completar los campos");
          return;
        }
        handlerAdd(form);
        setForm(initialDataForm);
      }}
    >
      <div>
        <input
          type="text"
          className="form-control my-3 w-75"
          name="Name"
          placeholder="name"
          value={name}
          onChange={(event) => setForm({ ...form, name: event.target.value })}
        />
      </div>
      <div>
        <input
          type="text"
          className="form-control my-3 w-75"
          name="Price"
          placeholder="price"
          value={price}
          onChange={(event) => setForm({ ...form, price: event.target.value })}
        />
      </div>
      <div>
        <input
          type="text"
          className="form-control my-3 w-75"
          name="Quantity"
          placeholder="quantity"
          value={quantity}
          onChange={(event) =>
            setForm({ ...form, quantity: event.target.value })
          }
        />
      </div>
      <div>
        <input
          type="text"
          className="form-control my-3 w-75"
          name="Descripcion"
          placeholder="description"
          value={description}
          onChange={(event) =>
            setForm({ ...form, description: event.target.value })
          }
        />
      </div>
      <div>
        <button type="submit" className="btn btn-primary">
          {id > 0 ? "Update" : "Create"}
        </button>
      </div>
    </form>
  );
};
