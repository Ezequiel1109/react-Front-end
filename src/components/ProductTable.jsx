import { ProductDetail } from "./ProductDetail"
import { PropTypes } from "prop-types"

export const ProductTable = ({handlerProductSelected, handlerRemove ,products = []}) =>{
    return(
        <table className="table table-hover table-striped">
            <thead>
                <tr>
                    <th>Nombre</th>
                    <th>Precio</th>
                    <th>Cantidad</th>
                    <th>Descripci√≥n</th>
                    <th>Update</th>
                    <th>Remove</th>
                </tr>
            </thead>
            <tbody>
                {products.map(product => {
                    return <ProductDetail handlerProductSelected={handlerProductSelected} handlerRemove={handlerRemove} product={product} key={product.id} />
                })}
            </tbody>
        </table>
    )
}

ProductTable.propTypes = {
    products: PropTypes.array.isRequired,
    handlerRemove: PropTypes.func.isRequired,
    handlerProductSelected: PropTypes.func.isRequired
}