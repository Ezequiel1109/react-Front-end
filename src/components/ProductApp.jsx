import { useEffect, useState } from "react";
import { create, findAll, remove, update} from "../services/ProductService";
import { ProductTable } from "./ProductTable";
import { PropTypes } from "prop-types";
import { ProductForm } from "./ProductForm";

export const ProductApp = ({ title }) => {
  const [products, setProducts] = useState([]);

  const [productSelected, setProductSelected] = useState({
    id: 0,
    name: '',
    price: '',
    quantity: '',
    description: ''
  })

  const getProducts = async () =>{
    const result = await findAll();
    setProducts(result.data._embedded.products);
  }

  useEffect(() => {
    getProducts();
  }, []);

  const handlerAddProduct = async (product) => {
    if (product.id > 0) {
      const response = await update(product);
      setProducts(products.map((prod) => {
          if (prod.id == response.data.id) {
            return { ...response.data };
          }
          return prod;
        })
      );
    } else {
      const response = await create(product);
      setProducts([...products, { ...response.data}]);
    }
  };
  const handlerRemoveProduct = (id) => {
    remove(id);
    setProducts(products.filter((product) => product.id != id));
  };

  const handlerProductSelected = (product) => {
    setProductSelected({ ...product });
  };

  return (
    <div className="container my-4">
      <h1>{title}</h1>
      <div className="row">
        <div className="col">
          <ProductForm
            handlerAdd={handlerAddProduct}
            productSelected={productSelected}
          />
        </div>
        <div className="col">{
            products.length > 0 ?
          <ProductTable
            products={products}
            handlerRemove={handlerRemoveProduct}
            handlerProductSelected={handlerProductSelected}
            />: <div className="alert alert-warning">No hay productos en el sistema!</div>
          }
        </div>
      </div>
    </div>
  );
};

ProductApp.propTypes = {
  title: PropTypes.string.isRequired,
}
