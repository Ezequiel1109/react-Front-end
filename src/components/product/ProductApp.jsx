import { useEffect, useState } from "react";
import { ProductTable } from "./ProductTable";
import { PropTypes } from "prop-types";
import { ProductForm } from "./ProductForm";
import { useSelector, useDispatch } from "react-redux";
import {
  addProduct,
  fetchProducts,
  updateProduct,
  deleteProduct,
} from "../../features/productsSlice";
import {
  selectIsAuth,
  selectProducts,
  selectProductsLoading,
} from "../../features/selectors";
import { Link } from "react-router-dom";

export const ProductApp = ({ title }) => {
  const dispatch = useDispatch();
  //usamos los selectores memoizados
  const products = useSelector(selectProducts);
  const isLoading = useSelector(selectProductsLoading);
  const isAuth = useSelector(selectIsAuth);

  const [message, setMessage] = useState(null);
  const [productSelected, setProductSelected] = useState({
    id: 0,
    name: "",
    price: "",
    quantity: "",
    description: "",
  });

  useEffect(() => {
    if (!isAuth) {
      dispatch(fetchProducts());
    }
    
  }, [dispatch, isAuth]);

  if (isAuth) {
    return (
      <div className="alert alert-warning text-center mt-5">
        <h4>Acceso Restringido</h4>
        <p>Por favor, <Link to="/login">inicia sesi√≥n</Link> para ver y gestionar los productos.</p>
      </div>
    )
  }

  const handlerAddProduct = async (product) => {
    if (
      !product.name ||
      !product.price ||
      !product.quantity ||
      !product.description ||
      isNaN(Number(product.price)) ||
      isNaN(Number(product.quantity))
    ) {
      setMessage({
        type: "danger",
        text: "Debe completar todos los campos correctamente",
      });
      return;
    }

    const productSend = {
      ...product,
      price: Number(product.price),
      quantity: Number(product.quantity),
    };
    try {
      if (product.id > 0) {
        await dispatch(updateProduct(productSend)).unwrap();
        setMessage({
          type: "success",
          text: "Producto actualizado correctamente",
        });
      } else {
        await dispatch(addProduct(productSend)).unwrap();
        setMessage({
          type: "success",
          text: "Producto agregado correctamente",
        });
      }
      setProductSelected({
        id: 0,
        name: "",
        price: "",
        quantity: "",
        description: "",
      });
    } catch (error) {
      setMessage({ type: "danger", text: error?.message || "Error al agregar el producto" });
    }
  };

  const handlerRemoveProduct = async (id) => {
    try {
      await dispatch(deleteProduct(id)).unwrap();
      setMessage({
        type: "success",
        text: "Producto eliminado correctamente",
      });
    } catch (error) {
      setMessage({ type: "danger", text: "Error al eliminar el producto" });
    }
  };

  const handlerProductSelected = (product) => {
    setProductSelected({ ...product });
  };

  return (
    <div className="container my-4">
      <h1>{title}</h1>
      {message && (
        <div className={`alert alert-${message.type}`}>
          {message.text}
          <button 
            type="button" 
            className="btn-close" 
            onClick={() => setMessage(null)}
          ></button>
        </div>
      )}
      <div className="row">
        <div className="col">
          <ProductForm
            handlerAdd={handlerAddProduct}
            productSelected={productSelected}
          />
        </div>
        <div className="col">
          {isLoading ? (
            <div className="d-flex justify-content-center mt-5">
              <div className="spinner-border" role="status">
                <span className="visually-hidden">Cargando...</span>
              </div>
            </div>
          ) : products.length > 0 ? (
            <ProductTable
              /* products={products}
              handlerRemove={handlerRemoveProduct}
              handlerProductSelected={handlerProductSelected} */
            />
          ) : (
            <div className="alert alert-warning">
              No hay productos en el sistema!
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

ProductApp.propTypes = {
  title: PropTypes.string.isRequired,
};
//export { ProductApp };
