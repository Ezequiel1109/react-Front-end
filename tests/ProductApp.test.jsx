import { render } from '@testing-library/react';
import { ProductApp } from '../src/components/ProductApp';

describe("Pruebas con el <ProductApp />", () => {
    it("deberia de renderizar correctamente", () => {
      const { getByText } = render(<ProductApp title={"Productos!"} />);
      expect(getByText("Productos!")).toMatchSnapshot();
    });
  });