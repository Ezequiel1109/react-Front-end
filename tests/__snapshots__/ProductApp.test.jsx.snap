// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con el <ProductApp /> deberia de renderizar correctamente 1`] = `
<h1>
  Productos!
</h1>
`;
